src/Type/ConditionalType.php

```diff
@@ -0,0 +1,97 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type;
+
+use PHPStan\Type\Generic\TemplateType;
+use PHPStan\Type\Generic\TemplateTypeVariance;
+use function sprintf;
+
+/** @api */
+final class ConditionalType extends UnionType
```

Comment:
An idea - do not implement UnionType, just Type, create `TypeCombinator::union($if, $else)` in the constructor and delegate all Type method calls to that type.

Reply:
Yeah, I think this is definitely a case where the classic principle of "prefer composition over inheritance" applies.

Reply:
Sure. For prototyping it was very useful though. :)

Reply:
Awesome! I love the progress, seems very promising :)